### PR TITLE
Add support for NeMo scope Optimizers support and add Novograd Optimizer

### PR DESCRIPTION
**Pull request by titu1994**
# Salient points

Refactor optimizer to support changing optimizers from command line. Run the codebase using the below examples

1) Default to Adam if no optimizer provided via `--optimizer`.
2) Create base optimizer if no overriding `--opt_args` are passed
3) Unify optimizer support for all domains of NeMo. Standardized interface to add optimizer, lr and opt_args to argparse.

### Use Adam and just override LR

```bash
python speech_to_text.py \
        --asr_model "bad_quartznet15x5.yaml" \
        --train_dataset "./an4/train_manifest.json" \
        --eval_dataset "./an4/test_manifest.json" \
        --gpus 4 \
        --distributed_backend "ddp" \
        --max_epochs 1 \
        --fast_dev_run \
        --lr 0.001 
```

### Change optimizer and override LR

```bash
python speech_to_text.py \
        --asr_model "bad_quartznet15x5.yaml" \
        --train_dataset "./an4/train_manifest.json" \
        --eval_dataset "./an4/test_manifest.json" \
        --gpus 4 \
        --distributed_backend "ddp" \
        --max_epochs 1 \
        --fast_dev_run \
        --optimizer novograd \
        --lr 0.01 
```

### Change optimizer, override LR, override optimizer args

```bash
python speech_to_text.py \
        --asr_model "bad_quartznet15x5.yaml" \
        --train_dataset "./an4/train_manifest.json" \
        --eval_dataset "./an4/test_manifest.json" \
        --gpus 4 \
        --distributed_backend "ddp" \
        --max_epochs 1 \
        --fast_dev_run \
        --optimizer novograd \
        --lr 0.01 \
        --opt_args betas=0.95,0.5 weight_decay=1e-3
```

## Usage - Overriding default args in the model itself

When calling `add_optimizer_args(parser)`, we can pass arguments directly here, thereby overriding the argparsers default values, as shown below. As such, with the same api, different domains can use different optimizers with different arguments.

### Base initialization

```python
parser = add_optimizer_args(parser)  # Use adam and empty opt_args list by default
```

###  Override optimizer

```python
parser = add_optimizer_args(parser, optimizer='novograd')  # Use novograd and empty opt_args list by default
```

### Override optimizer and args

```python
novograd_args = {betas:(0.95, 0.5), weight_decay:0.001}
parser = add_optimizer_args(parser, optimizer='novograd', default_opt_args=novograd_args)  # Use novograd and custom defaults
```

Signed-off-by: smajumdar <titu1994@gmail.com>

### CODE REVIEW

File: examples/asr/speech_to_text.py [hunk 1]
```diff
@@ -29,6 +30,7 @@
 
 from nemo.collections.asr.arguments import add_asr_args
 from nemo.collections.asr.models import EncDecCTCModel
+from nemo.core.classes.optimizers import add_optimizer_args
```
Comment (okuchaiev) [inline examples/asr/speech_to_text.py:33]: I think "optimizers" should be just under nemo.core.optimizers. Maybe rename that to nemo.core.optim since we can put LR policies there as well?

File: examples/asr/speech_to_text.py [hunk 2]
```diff
@@ -49,7 +51,7 @@ def main(args):
     model_config['AudioToTextDataLayer_eval']['manifest_filepath'] = args.eval_dataset
     asr_model.setup_training_data(model_config['AudioToTextDataLayer'])
     asr_model.setup_validation_data(model_config['AudioToTextDataLayer_eval'])
-    asr_model.setup_optimization(optim_params={'lr': 0.0003})
+    asr_model.setup_optimization(optim_params={'optimizer': args.optimizer, 'lr': args.lr, 'opt_args': args.opt_args})
```
Comment (okuchaiev) [inline examples/asr/speech_to_text.py:54]: so "lr" is outside or opt_args? Not sure about that

File: nemo/collections/asr/models/ctc_models.py [hunk 3]
```diff
@@ -79,7 +80,28 @@ def setup_test_data(self, test_data_layer_params: Optional[Dict]):
...
+        # We are guarenteed to have lr since it is required by the argparser
+        # But maybe user forgot to pass it to this function
+        lr = optim_params.get('lr', None)
+
+        if 'lr' is None:
+            raise ValueError('`lr` must be passed when setting up the optimization !')
+
+        # Actually instantiate the optimizer
+        optimizer = get_optimizer(optimizer_name)
+        self.__optimizer = optimizer(self.parameters(), lr=lr, **optimizer_args)
```
Comment (okuchaiev) [inline nemo/collections/asr/models/ctc_models.py:101]: is this where it will throw if wrong optimizer_args are passed? E.g. betas for SGD?

File: nemo/core/classes/optimizers.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+AVAILABLE_OPTIMIZERS = {
+    'sgd': optim.SGD,
+    'adam': optim.Adam,
+    'adamw': optim.AdamW,
+    'adadelta': adadelta.Adadelta,
+    'adamax': adamax.Adamax,
+    'adagrad': adagrad.Adagrad,
+    'rmsprop': rmsprop.RMSprop,
+    'rprop': rprop.Rprop,
```
Comment (okuchaiev) [inline nemo/core/classes/optimizers.py:35]: no "novograd"?

File: nemo/core/classes/optimizers.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+
+def _boolify(s):
```
Comment (okuchaiev) [inline nemo/core/classes/optimizers.py:39]: I'm not sure if functions like this should start with single _ or double __

File: nemo/core/classes/optimizers.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+    # If value is string `none` or `None`, parse as None
+    if value == 'none' or 'None':
+        return None
+
+    # Try type cast and return
+    for cast_type in (int, float, _boolify):
+        try:
+            return cast_type(value)
+        except Exception:
```
Comment (okuchaiev) [inline nemo/core/classes/optimizers.py:66]: shouldn't we fail in this case?

File: nemo/core/classes/optimizers.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    Generator expression that iterates over the params owned by ``optimizer``.
+    Args:
+        optimizer: An optimizer previously returned from ``amp.initialize``.
+    """
+    for group in optimizer.param_groups:
+        for p in group['params']:
+            yield p
+
+
+class Novograd(Optimizer):
```
Comment (okuchaiev) [inline nemo/core/classes/optimizers.py:162]: please move Novograd optimizer into separate class

File: nemo/core/classes/optimizers.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (okuchaiev) [inline nemo/core/classes/optimizers.py:1]: let's put this under nemo.core.optim.Novograd.py and nemo.core.optim.optimizers.py

__all__  accessible from nemo.core.optim.*

File: nemo/core/classes/optimizers.py (hunk 6 above)
Comment (titu1994) [reply nemo/core/classes/optimizers.py:66]: No, it will try all possibly casts - in the order of dtype - int then float then bool. Above it tests for None. If it escapes everything, its a string type value.

File: nemo/core/classes/optimizers.py (hunk 7 above)
Comment (titu1994) [reply nemo/core/classes/optimizers.py:162]: Sounds good. Under `nemo.core.optim.novograd`?

File: examples/asr/speech_to_text.py (hunk 2 above)
Comment (okuchaiev) [reply examples/asr/speech_to_text.py:54]: sounds good

File: nemo/core/classes/optimizers.py (hunk 7 above)
Comment (okuchaiev) [reply nemo/core/classes/optimizers.py:162]: yeah, something like nemo/core/optim/Novograd.py

File: nemo/collections/asr/models/ctc_models.py [hunk 9]
```diff
@@ -79,7 +80,28 @@ def setup_test_data(self, test_data_layer_params: Optional[Dict]):
         self.__test_dl = self.__setup_dataloader_from_config(config=test_data_layer_params)
 
     def setup_optimization(self, optim_params: Optional[Dict]):
```
Comment (blisc) [inline nemo/collections/asr/models/ctc_models.py:82]: optim_params is described to be optional but has no default parameter

File: examples/asr/speech_to_text.py (hunk 2 above)
Comment (blisc) [inline examples/asr/speech_to_text.py:54]: I suppose this is out of scope of PR, but these three lines look holly out of line with pytorch lightning code. Just do all of this in init(). I fail to see the reason we need to do this separately.

File: nemo/collections/asr/models/ctc_models.py (hunk 3 above)
Comment (blisc) [inline nemo/collections/asr/models/ctc_models.py:101]: why not merge these two lines into one?

File: nemo/collections/asr/models/ctc_models.py [hunk 10]
```diff
@@ -79,7 +80,28 @@ def setup_test_data(self, test_data_layer_params: Optional[Dict]):
...
+
+        if 'lr' is None:
+            raise ValueError('`lr` must be passed when setting up the optimization !')
+
+        # Actually instantiate the optimizer
+        optimizer = get_optimizer(optimizer_name)
+        self.__optimizer = optimizer(self.parameters(), lr=lr, **optimizer_args)
+
+        # TODO: Remove after demonstration
+        logging.info("Optimizer config = %s", str(self.__optimizer))
```
Comment (blisc) [inline nemo/collections/asr/models/ctc_models.py:104]: Why have this function as part of ctc_models? Why not add this functionality to core?

File: nemo/core/optim/optimizers.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    parser.add_argument('--optimizer', type=str, default=optimizer, help='Name of the optimizer. Defaults to Adam.')
+    parser.add_argument('--lr', type=float, required=True, help='Learning rate of the optimizer.')
+    parser.add_argument(
+        '--opt_args',
+        default=default_opt_args,
+        nargs='+',
+        type=str,
+        help='Overriding arguments for the optimizer. \n'
+        'Must follow the pattern : \n'
+        'name=value seperated by spaces.',
```
Comment (blisc) [inline nemo/core/optim/optimizers.py:119]: This help can be merged into less lines, can you explanation how to pass a list using this?

File: nemo/core/optim/optimizers.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+from argparse import ArgumentParser
+from functools import partial
+
+import torch.optim as optim
+from torch.optim import adadelta, adagrad, adamax, rmsprop, rprop
+from torch.optim.optimizer import Optimizer
+
+from nemo.core.optim.novograd import Novograd
+
+__all__ = ['get_optimizer', 'register_optimizer', 'parse_optimizer_args', 'add_optimizer_args']
```
Comment (blisc) [inline nemo/core/optim/optimizers.py:24]: Please add a docstring to each of these functions.

File: nemo/core/optim/optimizers.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+def master_params(optimizer):
+    """
+    Generator expression that iterates over the params owned by ``optimizer``.
+    Args:
+        optimizer: An optimizer previously returned from ``amp.initialize``.
+    """
+    for group in optimizer.param_groups:
+        for p in group['params']:
+            yield p
```
Comment (blisc) [inline nemo/core/optim/optimizers.py:151]: Seems unused, can you remove?

**Comment by blisc** [comment]
Seems mostly fine to me, just some minor comments

File: examples/asr/speech_to_text.py (hunk 2 above)
Comment (okuchaiev) [reply examples/asr/speech_to_text.py:54]: @blisc are you proposing to have models.__init__() take: (1) model hyper parameters, (2) optimizer hyper parameters and (3) train/test/eval data parameters instead of having setup_* functions?

File: nemo/core/optim/optimizers.py (hunk 11 above)
Comment (titu1994) [reply nemo/core/optim/optimizers.py:119]: An explanation of how to pass lists inside the help? The help section will get kinda long, but I can try summarizing it succinctly.

File: nemo/core/optim/optimizers.py (hunk 13 above)
Comment (titu1994) [reply nemo/core/optim/optimizers.py:151]: Oh its from old codebase. Good catch.

File: examples/asr/speech_to_text.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply examples/asr/speech_to_text.py:54]: @blisc I came to exactly the same conclusions yesterday - thus my email.

I think the solution is to properly _parametrize_ NeMo Models.

Comment (titu1994) [reply examples/asr/speech_to_text.py]: We no longer need to manually extract the kwargs from the parsed args, `vars(args)` is concise and serves the same purpose.

File: examples/asr/speech_to_text.py [hunk 14]
```diff
@@ -49,7 +51,7 @@ def main(args):
     model_config['AudioToTextDataLayer_eval']['manifest_filepath'] = args.eval_dataset
     asr_model.setup_training_data(model_config['AudioToTextDataLayer'])
     asr_model.setup_validation_data(model_config['AudioToTextDataLayer_eval'])
-    asr_model.setup_optimization(optim_params={'lr': 0.0003})
+    asr_model.setup_optimization(optim_params=vars(args))
```
Comment (VahidooX) [inline examples/asr/speech_to_text.py:54]: Are we passing all args here?I If yes, I think it does not look nice that we pass all the args here. Can we pass three variables: lr, optimizer_kind and opt_params? like asr_model.setup_optimization(lr=args.lr, optimizer_kind=args.optimizer_kind, optim_params=args.opt_param)?

It makes it easier to understandable by the user. I think it looks like a magic function like this that gets all the args and return the optimizer?

File: nemo/collections/asr/models/ctc_models.py [hunk 15]
```diff
@@ -78,8 +77,8 @@ def setup_test_data(self, test_data_layer_params: Optional[Dict]):
             test_data_layer_params['shuffle'] = False
         self.__test_dl = self.__setup_dataloader_from_config(config=test_data_layer_params)
 
-    def setup_optimization(self, optim_params: Optional[Dict]):
-        self.__optimizer = torch.optim.Adam(self.parameters(), lr=optim_params['lr'])
+    def setup_optimization(self, optim_params: Optional[Dict] = None) -> torch.optim.Optimizer:
```
Comment (VahidooX) [inline nemo/collections/asr/models/ctc_models.py:80]: Do we need to have setup_optimization here if it is written already in the higher clas?

File: examples/asr/speech_to_text.py (hunk 14 above)
Comment (titu1994) [reply examples/asr/speech_to_text.py:54]: Yes, that is what it was before, and we can go back to that. I'm a bit worried about how many args are going to be passed for optimizer + scheduler but might be better to be explicit.

File: nemo/collections/asr/models/ctc_models.py (hunk 15 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_models.py:80]: I added if because it's needed for a later PR of schedulers

File: examples/asr/speech_to_text.py (hunk 14 above)
Comment (titu1994) [reply examples/asr/speech_to_text.py]: Reverted it to pass args explicitly. I agree, this looks cleaner and more understandable.

Comment (VahidooX) [reply examples/asr/speech_to_text.py]: It is a good point about the scheduler. How about having 5 inputs: 1- optimizer_kind, 2-lr 3-opt_params 4-lr_policy 5-lr_policy_params?

Comment (titu1994) [reply examples/asr/speech_to_text.py]: That's a good idea. We should be able to do that cleanly. Though for now, I am hard coding scheduler until the team finds this approach to Optimizer good enough to extend to schedulers as well.

File: nemo/collections/asr/models/ctc_models.py (hunk 3 above)
Comment (blisc) [reply nemo/collections/asr/models/ctc_models.py]: This wasn't what I had in mind actually. I was thinking more `return get_optimizer(optimizer_name, self.parameters(), lr=lr, **optimizer_args)`, ie I would expect get_optimizer to instantiate an optimizer for me.

If you want to keep your original design, I would actually prefer the old:
```
optimizer = get_optimizer(optimizer_name)
self.__optimizer = optimizer(self.parameters(), lr=lr, **optimizer_args)
```
rather than the changed:
```
optimizer = get_optimizer(optimizer_name)(self.parameters(), lr=lr, **optimizer_args)
```

Comment (titu1994) [reply nemo/collections/asr/models/ctc_models.py]: Oh I misunderstood. Yes, I'll revert to follow the older design. As to merging the two lines together, I would prefer not to do that for two reasons - 1) we may want the class without instantiation to wrap into another class (say we have experimental optimizer), 2) we want to pass the class as an argument without instantiation to perform defered computation or typecheck in tests.
